### PR TITLE
Rewrite the plugin in ES2015, switch to catberry-locator v2

### PR DESCRIPTION
**Pull request by reenko**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -8,7 +8,7 @@ TESTS = test/lib/* \
 all: lint test
```
Comment (pragmadash) [inline Makefile]: We need to copy the whole Makefile from https://github.com/catberry/catberry/blob/develop/Makefile and replace `SOURCES` and `TESTS` values.

File: browser/LocalizationLoader.js [hunk 2]
```diff
@@ -30,29 +30,26 @@
...
-		typeof ($window.localization) === 'object' ? $window.localization : {};
-}
+	/**
+	 * Creates new instance of localization loader.
+	 * @param {ServiceLocator} locator Locator to resolve dependencies.
+	 * @constructor
+	 */
+	constructor(locator) {
+		const window = locator.resolve('window');
+		this._localization = window.localization &&
```
Comment (pragmadash) [inline browser/LocalizationLoader.js]: Lost jsDoc for the property.

File: browser/LocalizationLoader.js [hunk 3]
```diff
@@ -30,29 +30,26 @@
...
+		typeof (window.localization) === 'object' ? window.localization : {};
```
Comment (pragmadash) [inline browser/LocalizationLoader.js]: Need to add an indentation here.

File: browser/LocalizationLoader.js [hunk 4]
```diff
@@ -30,29 +30,26 @@
...
- * @constructor
- */
-function LocalizationLoader($window) {
-	this._localization = $window.localization &&
-		typeof ($window.localization) === 'object' ? $window.localization : {};
-}
+	/**
+	 * Creates new instance of localization loader.
+	 * @param {ServiceLocator} locator Locator to resolve dependencies.
+	 * @constructor
```
Comment (pragmadash) [inline browser/LocalizationLoader.js]: We don't need a `@constructor` annotation anymore, it's obvious that the method is a constructor.

File: lib/LocalizationLoader.js [hunk 5]
```diff
@@ -30,479 +30,393 @@
...
+const LOCALE_URL = '/l10n.js';
+
+/* eslint max-nested-callbacks: [2, 4]*/
+
+class LocalizationLoader {
+
+	/**
+	 * Creates new instance of server-side localization loader.
+	 * @param {ServiceLocator} locator Locator to resolve dependencies.
+	 * @constructor
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: Remove the annotation.

File: lib/LocalizationLoader.js [hunk 6]
```diff
@@ -30,479 +30,393 @@
...
+	 */
+	constructor(locator) {
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: Lost all jsDocs for the properties.

File: lib/LocalizationLoader.js [hunk 7]
```diff
@@ -30,479 +30,393 @@
...
-		path.join(process.cwd(), LOCALIZATIONS_FOLDER_NAME));
-	this._componentFinder = $serviceLocator.resolve('componentFinder');
-	this._logger = $serviceLocator.resolve('logger');
-
-	this._defaultLocale = l10n.defaultLocale ?
-		String(l10n.defaultLocale) : '';
-	if (!LOCALE_REGEXP.test(this._defaultLocale)) {
-		throw new Error(util.format(ERROR_LOCALE_NAME, this._defaultLocale));
+		this._eventBus = locator.resolve('eventBus');
+		this._eventBus.on('allComponentsLoaded', () => {
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: `this._eventBus.on('allComponentsLoaded', () => this._init());`

File: lib/LocalizationLoader.js [hunk 8]
```diff
@@ -30,479 +30,393 @@
...
-module.exports = LocalizationLoader;
+const path = require('path');
+const fs = require('./promises/fs');
+const url = require('url');
+const pluralizationRules = require('./pluralizationRules.json');
+
+const LOCALIZATIONS_FOLDER_NAME = 'l10n';
+const BROWSER_LOCALE_REGEXP = /[a-z]{2}(-[a-z]{2})?/i;
+const LOCALE_REGEXP = /^[a-z]{2}(-[a-z]{2})?$/;
+const ERROR_LOCALIZATION_CONFIG = '"l10n" config section is required';
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: We can use it once in the constructor and remove it from here.

File: lib/LocalizationLoader.js [hunk 9]
```diff
@@ -30,479 +30,393 @@
...
+				const fileDirectory = path.dirname(args.filename);
+				if (fileDirectory !== l10nDirectory) {
+					return;
+				}
+				delete this._loadingCache[args.filename];
+				this._init();
+			})
+			.on('add', () => {
+				this._init();
+			})
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: `.on('add', () => this._init())`

File: lib/LocalizationLoader.js [hunk 10]
```diff
@@ -30,479 +30,393 @@
...
+	_init() {
+		return this._prepareCache()
+			.then(() => {
+				const defaultL10n = this._objectCache[this._defaultLocale];
+				if (!defaultL10n || typeof (defaultL10n) !== 'object') {
+					throw new Error(`Can not load default locale ${this._defaultLocale}`);
+				}
+			})
+			.then(() => {
+				this._eventBus.emit('l10nLoaded');
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: Remove `{}`.

File: lib/LocalizationLoader.js [hunk 11]
```diff
@@ -30,479 +30,393 @@
...
+			})
+			.catch(reason => {
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: Remove `{}`.

File: lib/LocalizationLoader.js [hunk 12]
```diff
@@ -30,479 +30,393 @@
...
-		typeof (this._fileCache[firstLocaleComponent]) === 'string') {
-		return this._fileCache[firstLocaleComponent];
+	/**
+	 * Initializes pluralization map.
+	 * @private
+	 */
+	_initPluralizationMap() {
+		const pluralizationMap = this._pluralizationMap = {};
+		Object.keys(pluralizationRules)
+			.forEach(rule => {
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: Remove `{}`.

File: lib/LocalizationLoader.js [hunk 13]
```diff
@@ -30,479 +30,393 @@
...
+						if (localizationsStat.isDirectory()) {
+							// load basic localization
+							return this._loadLocalizationsFromPath(
+								localizations, this._localizationsPath);
+						}
+						return null;
+					});
+			})
+			.then(() => {
+				return this._componentFinder.find();
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: Remove `{}`.

File: lib/LocalizationLoader.js [hunk 14]
```diff
@@ -30,479 +30,393 @@
...
- * @param {string} loadPath Path where localization files is there.
- * @returns {Promise} Promise for operations is complete.
- * @private
- */
-LocalizationLoader.prototype._loadLocalizationsFromPath =
-	function (localizations, loadPath) {
-		var self = this;
+				return Promise.all(promises);
+			})
+			.then(() => {
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: Remove `{}`.

File: lib/LocalizationLoader.js [hunk 15]
```diff
@@ -511,74 +425,70 @@ LocalizationLoader.prototype._loadLocalizationsFromPath =
...
 		if (this._cookieConfig.path) {
-			cookieSetup += '; Path=' + String(this._cookieConfig.path || '');
+			cookieSetup += `; Path=${String(this._cookieConfig.path || '')}`;
 		}
 
 		if (this._cookieConfig.domain) {
-			cookieSetup +=
-				'; Domain=' + String(this._cookieConfig.domain || '');
+			cookieSetup += `; Domain=${String(this._cookieConfig.domain || '')}`;
 		}
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: Please check if it's up to date with https://github.com/catberry/catberry/blob/develop/lib/base/CookieWrapperBase.js#L86-L111

File: lib/LocalizationProvider.js [hunk 16]
```diff
@@ -1,214 +1,184 @@
...
+* OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+* SOFTWARE.
+*
+* This license applies to all parts of catberry that are not externally
+* maintained libraries.
+*/
```
Comment (pragmadash) [inline lib/LocalizationProvider.js]: There is no need to keep the license header.

File: lib/LocalizationProvider.js [hunk 17]
```diff
@@ -1,214 +1,184 @@
...
-function LocalizationProvider($localizationLoader, l10n) {
-	if (!l10n) {
-		throw new Error(ERROR_LOCALIZATION_CONFIG);
-	}
-	this._defaultLocale = l10n.defaultLocale ?
+const DEFAULT_LOCALE_COOKIE_KEY = 'locale';
+const DEFAULT_LOCALE_COOKIE_MAX_AGE = 3155692600; // 100 years
+const LOCALE_COOKIE_PATH = '/';
+const LOCALE_REGEXP = /^[a-z]{2}(-[a-z]{2})?$/;
+const ERROR_LOCALIZATION_CONFIG = '"l10n" config section is required';
```
Comment (pragmadash) [inline lib/LocalizationProvider.js]: Remove from here and use once in the constructor.

File: lib/LocalizationProvider.js [hunk 18]
```diff
@@ -1,214 +1,184 @@
...
+
+	/**
+	 * Gets localized value for specified locale and key name.
+	 * @param {string} locale Locale name (i.e. EN, RU etc).
+	 * @param {string} key Localization key.
+	 * @returns {string} Localized value.
+	 */
+	get(locale, key) {
+		let value = this._loader.load(locale)[key];
+		if (value instanceof Array) {
```
Comment (pragmadash) [inline lib/LocalizationProvider.js]: I would use `Array.isArray` here too.

File: lib/LocalizationProvider.js [hunk 19]
```diff
@@ -1,214 +1,184 @@
...
+	 */
+	pluralize(locale, key, n) {
+		const localeObject = this._loader.load(locale);
+		const forms = localeObject[key];
+
+		if (!forms && this._placeholder) {
+			return key;
+		}
+
+		if (!(forms instanceof Array)) {
```
Comment (pragmadash) [inline lib/LocalizationProvider.js]: I would use `Array.isArray` as well.

File: test/browser/LocalizationLoader.js [hunk 20]
```diff
@@ -30,29 +30,30 @@
...
-		it('should return the same object from window on all specified locales',
-			function () {
-				var locator = new ServiceLocator(),
-					localization = {
-						TEST_VALUE: 'test'
-					};
+/* eslint max-nested-callbacks: [2, 4]*/
+
+describe('client/LocalizationLoader', () => {
+	describe('#load', () => {
```
Comment (pragmadash) [inline test/browser/LocalizationLoader.js]: Hmm, interesting. Mocha maintainers say [that it does not work with arrow functions](https://mochajs.org/#arrow-functions). Are you sure that works?

File: test/lib/LocalizationLoader.js [hunk 21]
```diff
@@ -95,80 +93,88 @@ localizations.ru.$pluralization = {
...
-					LocalizationLoader, defaultConfig
-				);
+	describe('#load', () => {
+		it('should throw exception on wrong locale', () => {
+			const locator = createLocator();
+			locator.registerInstance('config', defaultConfig);
+			const localizationLoader = new LocalizationLoader(locator);
 
-			assert.throws(function () {
+			assert.throws(() => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 22]
```diff
@@ -351,7 +353,7 @@ describe('server/LocalizationLoader', function () {
 								'Response cookie should have locale'
 							);
 
-							server.close(function () {
+							server.close(() => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 23]
```diff
@@ -406,38 +405,38 @@ describe('server/LocalizationLoader', function () {
...
+								`testName=en-us\
+; Max-Age=${config.l10n.cookie.maxAge}\
+; Expires=${expireDate.toUTCString()}\
+; Path=${config.l10n.cookie.path}\
+; Domain=${config.l10n.cookie.domain}`,
 								'Response cookie should have locale'
 							);
 
-							server.close(function () {
+							server.close(() => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 24]
```diff
@@ -406,38 +405,38 @@ describe('server/LocalizationLoader', function () {
...
 								done();
 							});
 						});
 
 					expireDate = new Date((new Date()).getTime() +
-					config.cookie.maxAge * 1000);
+					config.l10n.cookie.maxAge * 1000);
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Nice catch, you've fixed the bug. Is covered by any test right now?

File: test/lib/LocalizationLoader.js [hunk 25]
```diff
@@ -453,13 +452,14 @@ describe('server/LocalizationLoader', function () {
 								'Response should have one cookie setup'
 							);
 
-							assert.strictEqual(/^locale=ru/
+							// TODO: refactoring, old value 'locale'
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Is it here by mistake or we still need to do something?

File: test/lib/LocalizationLoader.js [hunk 26]
```diff
@@ -453,13 +452,14 @@ describe('server/LocalizationLoader', function () {
...
+							assert.strictEqual(/^testName=ru/
 									.test(response.headers['set-cookie'][0]),
 								true,
 								'Response cookie should have locale'
 							);
 
-							server.close(function () {
+							server.close(() => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 27]
```diff
@@ -491,7 +491,7 @@ describe('server/LocalizationLoader', function () {
 								undefined,
 								'Response should not have cookies');
 
-							server.close(function () {
+							server.close(() => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 28]
```diff
@@ -526,59 +525,59 @@ describe('server/LocalizationLoader', function () {
...
+										/* eslint no-eval: 0*/
+										const window = {};
 										eval(data);
 										assert.deepEqual(
 											window.localization,
 											mergedLocalization,
 											'Localization do not match'
 										);
-										server.close(function () {
+										server.close(() => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 29]
```diff
@@ -526,59 +525,59 @@ describe('server/LocalizationLoader', function () {
...
 											done();
 										});
 									});
 							});
 						request.end();
 					});
-				server.listen(8084, function () {
+				server.listen(8084, () => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 30]
```diff
@@ -590,110 +589,116 @@ describe('server/LocalizationLoader', function () {
...
+										/* eslint no-eval: 0*/
+										const window = {};
 										eval(data);
 										assert.deepEqual(
 											window.localization,
 											mergedLocalization,
 											'Localization do not match'
 										);
-										server.close(function () {
+										server.close(() => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 31]
```diff
@@ -590,110 +589,116 @@ describe('server/LocalizationLoader', function () {
...
 											done();
 										});
 									});
 							});
 						request.end();
 					});
-				server.listen(8085, function () {
+				server.listen(8085, () => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 32]
```diff
@@ -590,110 +589,116 @@ describe('server/LocalizationLoader', function () {
...
+										/* eslint no-eval: 0*/
+										const window = {};
 										eval(data);
 										assert.deepEqual(
 											window.localization,
 											defaultLocalization,
 											'Localization do not match'
 										);
-										server.close(function () {
+										server.close(() => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 33]
```diff
@@ -590,110 +589,116 @@ describe('server/LocalizationLoader', function () {
...
 											done();
 										});
 									});
 							});
 						request.end();
 					});
-				server.listen(8086, function () {
+				server.listen(8086, () => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationLoader.js [hunk 34]
```diff
@@ -590,110 +589,116 @@ describe('server/LocalizationLoader', function () {
...
 function createLocator(components) {
 	components = components || {};
-	var locator = new ServiceLocator();
+	const locator = new ServiceLocator();
 	locator.register('logger', Logger);
 	locator.registerInstance('serviceLocator', locator);
-	var componentFinder = new events.EventEmitter();
-	componentFinder.find = function () {
+	const componentFinder = new events.EventEmitter();
+	componentFinder.find = () => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}` and `return`.

File: test/lib/LocalizationLoader.js [hunk 35]
```diff
@@ -701,12 +706,18 @@ function createLocator(components) {
...
+ * @param middleware
+ * @param endCallback
+ * @returns {*}
+ */
 function createServer(middleware, endCallback) {
-	endCallback = endCallback || function () {};
-	return http.createServer(function (request, response) {
-		middleware(request, response, function () {
+	endCallback = endCallback || function() {};
+	return http.createServer((request, response) => {
```
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: Remove `{}`.

File: test/lib/LocalizationProvider.js [hunk 36]
```diff
@@ -453,19 +432,23 @@ describe('LocalizationProvider', function () {
...
 function createLocator(config) {
-	var locator = new ServiceLocator();
+	const locator = new ServiceLocator();
 	locator.registerInstance('config', config);
-	locator.registerInstance('serviceLocator', locator);
 	locator.registerInstance('eventBus', new events.EventEmitter());
-	var componentFinder = new events.EventEmitter();
-	componentFinder.find = function () {
+	const componentFinder = new events.EventEmitter();
+	componentFinder.find = () => {
```
Comment (pragmadash) [inline test/lib/LocalizationProvider.js]: Remove `{}` and `return`.

**Comment by pragmadash** [comment]
It also would be nice to refactor tests with HTTP server the same way as it's done with [UHR tests](https://github.com/catberry/catberry-uhr/blob/develop/test/lib/UHR.js#L17-L25). We would remove many lines of code then.

File: test/lib/LocalizationLoader.js (hunk 25 above)
Comment (reenko) [inline test/lib/LocalizationLoader.js]: I change `assert.strictEqual(/^locate=ru/` to `assert.strictEqual(/^testName=ru/`for correct test and I doubt it.

File: test/lib/LocalizationLoader.js (hunk 24 above)
Comment (reenko) [inline test/lib/LocalizationLoader.js]: Yes, in all test use `config.l10n.cookie`

File: Makefile (hunk 1 above)
Comment (reenko) [inline Makefile]: Done.

File: test/browser/LocalizationLoader.js (hunk 20 above)
Comment (reenko) [inline test/browser/LocalizationLoader.js]: I think, yes :)

File: lib/LocalizationLoader.js [hunk 37]
```diff
@@ -30,555 +30,467 @@
...
+	/**
+	 * Creates new instance of server-side localization loader.
+	 * @param {ServiceLocator} locator Locator to resolve dependencies.
+	 */
+	constructor(locator) {
+		const l10n = locator.resolve('config').l10n;
+		if (!l10n) {
+			throw new Error('"l10n" config section is required');
+		}
+		this._localizationsPath = String(l10n.path ||
```
Comment (pragmadash) [inline lib/LocalizationLoader.js]: There is still no jsDoc for all the properties defined in the constructor.

File: lib/LocalizationProvider.js [hunk 38]
```diff
@@ -30,185 +30,154 @@
...
-		throw new Error(ERROR_LOCALIZATION_CONFIG);
-	}
-	this._defaultLocale = l10n.defaultLocale ?
+class LocalizationProvider {
+
+	/**
+	 * Creates new instance of localization provider.
+	 * @param {ServiceLocator} locator Locator to resolve dependencies.
+	 */
+	constructor(locator) {
```
Comment (pragmadash) [inline lib/LocalizationProvider.js]: Keep jsDoc for every property defined inside the constructor.

File: test/lib/LocalizationLoader.js (hunk 25 above)
Comment (pragmadash) [inline test/lib/LocalizationLoader.js]: So, if I got it right it's for cookie name parameter testing? If so, just remove the todo comment.

**Pull request merged**